kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py

```diff
@@ -0,0 +1,189 @@
+from __future__ import annotations
+
+import importlib
+from copy import deepcopy
+from pathlib import PurePosixPath
+from typing import Any
+
+import fsspec
+from kedro.io.core import (
+    AbstractVersionedDataset,
+    DatasetError,
+    Version,
+    get_filepath_str,
+    get_protocol_and_path,
+)
+
+
+class SafetensorsDataset(AbstractVersionedDataset[Any, Any]):
+    """``SafetensorsDataset`` loads/saves data from/to a Safetensors file using an underlying
+    filesystem (e.g.: local, S3, GCS). The underlying functionality is supported by
+    the specified backend library passed in (defaults to the ``torch`` library), so it
+    supports all allowed options for loading and Safetensors files.
+
+    Example usage for the
+    `YAML API <https://docs.kedro.org/en/stable/data/\
+    data_catalog_yaml_examples.html>`_:
+
+    .. code-block:: yaml
+
+        test_model: # simple example without compression
+          type: safetensors.SafetensorsDataset
+          filepath: data/07_model_output/test_model.safetensors
+          backend: torch
+
+    Example usage for the
+    `Python API <https://docs.kedro.org/en/stable/data/\
+    advanced_data_catalog_usage.html>`_:
+
+    .. code-block:: pycon
+
+        >>> from kedro_datasets.safetensors import SafetensorsDataset
+        >>> import torch
+        >>>
+        >>> data = {"embeddings": torch.zeros((10, 100)}
+        >>> dataset = SafetensorsDataset(
+        ...     filepath="test.safetensors",
+        ...     backend="torch"
+        ... )
+        >>> dataset.save(data)
+        >>> reloaded = dataset.load()
+        >>> assert data.equals(reloaded)
+    """
+
+    DEFAULT_LOAD_ARGS: dict[str, Any] = {}
+    DEFAULT_SAVE_ARGS: dict[str, Any] = {}
+    DEFAULT_FS_ARGS: dict[str, Any] = {"open_args_save": {"mode": "wb"}}
+
+    def __init__(  # noqa: PLR0913
+        self,
+        *,
+        filepath: str,
+        backend: str = "torch",
```

Comment:
PyTorch is notably tricky to install, should we default to `numpy`? It's also the first backend shown in the docs https://pypi.org/project/safetensors/

Reply:
And if we're choosing a default backend, should we add that to the dependencies? I think it's not too high stakes to depend on NumPy for this dataset.

Reply:
Even the PyTorch backend seems to depend on NumPy

```
In [2]: ds = SafetensorsDataset(filepath="/tmp/output.safetensors", backend="torch")
/Users/juan_cano/.cache/uv/archive-v0/SRNuoVa2WtZC4Q79225OL/lib/python3.12/site-packages/torch/_subclasses/functional_tensor.py:295: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/torch/csrc/utils/tensor_numpy.cpp:84.)
  cpu = _conversion_method_template(device=torch.device("cpu"))
```

Reply:
@astrojuanlu That makes sense. I've updated the default backend to be `numpy` and added it to the dependencies. 
Do you want me to update the examples in the docstring to be `numpy` as well or should we leave them in `torch`?

Reply:
Better to update the docstrings yes! @MinuraPunchihewa

Reply:
@astrojuanlu It's done!